be/src/olap/rowset/segment_v2/indexed_column_reader.cpp

```diff
@@ -17,55 +17,254 @@
+    if (_verify_checksum) {
+        uint32_t expect = decode_fixed32_le((uint8_t*)page_slice.data + page_slice.size - 4);
+        uint32_t actual = crc32c::Value(page_slice.data, page_slice.size - 4);
+        if (expect != actual) {
+            return Status::Corruption(
+                Substitute("Page checksum mismatch, actual=$0 vs expect=$1", actual, expect));
+        }
+    }
+
+    // remove page's suffix
+    page_slice.size = data_size;
+    if (_compress_codec != nullptr) {
+        PageDecompressor decompressor(page_slice, _compress_codec);
+
+        Slice uncompressed_page;
+        RETURN_IF_ERROR(decompressor.decompress_to(&uncompressed_page));
+
+        // If decompressor create new heap memory for uncompressed data,
+        // assign this uncompressed page to page and page slice
+        if (uncompressed_page.data != page_slice.data) {
+            page.reset((uint8_t*)uncompressed_page.data);
+        }
+        page_slice = uncompressed_page;
+    }
+    // insert this into cache and return the cache handle
+    cache->insert(cache_key, page_slice, &cache_handle);
+    page.release();
+    *handle = PageHandle(std::move(cache_handle));
+
+    return Status::OK();
 }
 
 ///////////////////////////////////////////////////////////////////////////////
 
+Status IndexedColumnIterator::_read_data_page(const PagePointer& page_pointer, ParsedPage* page) {
+    RETURN_IF_ERROR(_reader->read_page(page_pointer, &page->page_handle));
+    Slice data = page->page_handle.data();
 
-IndexedColumnIterator::IndexedColumnIterator(IndexedColumnReader* reader) {
-    // TODO
-}
-IndexedColumnIterator::~IndexedColumnIterator() {
-    // TODO
-}
-Status IndexedColumnIterator::seek_to_first() {
-    return Status(); // TODO
+    // decode first rowid
+    if (!get_varint32(&data, &page->first_rowid)) {
+        return Status::Corruption("Bad page, failed to decode first rowid");
+    }
+
+    // decode number rows
+    if (!get_varint32(&data, &page->num_rows)) {
+        return Status::Corruption("Bad page, failed to decode rows count");
+    }
+
+    // create page data decoder
+    PageDecoderOptions options;
+    RETURN_IF_ERROR(_reader->encoding_info()->create_page_decoder(data, options, &page->data_decoder));
+    RETURN_IF_ERROR(page->data_decoder->init());
+
+    page->offset_in_page = 0;
+    return Status::OK();
 }
+
 Status IndexedColumnIterator::seek_to_ordinal(rowid_t idx) {
-    return Status(); // TODO
-}
-Status IndexedColumnIterator::seek_at_or_after(const Slice& key, bool* exact_match) {
-    return Status(); // TODO
+    DCHECK(idx >= 0 && idx <= _reader->num_values());
+
+    if (!_reader->support_ordinal_seek()) {
+        return Status::NotSupported("no ordinal index");
+    }
+
+    // it's ok to seek past the last value
+    if (idx == _reader->num_values()) {
+        _current_rowid = idx;
+        _seeked = true;
+        return Status::OK();
+    }
+
+    if (!_data_page || !_data_page->contains(idx)) {
```

Comment:
```suggestion
    if (_data_page != nullptr || !_data_page->contains(idx)) {
```
Only use `!` for bool type.

Reply:
OK. we firstly check is `_data_page == nullptr`, not `_data_page != nullptr`